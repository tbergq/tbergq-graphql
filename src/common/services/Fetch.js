// @flow

import fetch from 'node-fetch';

export default async (url: string, options: Object = {}) => {
  try {
    const response = await fetch(url, options);

    const json = await response.json();
    return json;
  } catch (err) {
    console.log(err);
    throw err;
  }
};
