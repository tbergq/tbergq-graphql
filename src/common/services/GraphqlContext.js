// @flow

import DataLoader from 'dataloader';
import jwt from 'jsonwebtoken';
import { config } from 'dotenv';

import SearchTvShowLoader, {
  type TvShow,
} from '../../tvhelper/dataloaders/SearchTvShowLoader';
import TvDetailLoader from '../../tvhelper/dataloaders/TvDetailLoader';
import TvHelperUserLoader, {
  type UserType as TvHelperUser,
} from '../../tvhelper/dataloaders/UserLoader';
import FavoritesLoader, {
  type Favorites,
} from '../../tvhelper/dataloaders/FavoritesLoader';
import TvShowEpisodesLoader, {
  type Episode,
} from '../../tvhelper/dataloaders/TvShowEpisodesLoader';
import FavoriteLoader, {
  type FavoriteArgs,
} from '../../tvhelper/dataloaders/FavoriteLoader';
import EpiosodeWatchedLoader, {
  type EpisodeWatched,
} from '../../tvhelper/dataloaders/EpisodeWatchedLoader';
import EpisodeLoader from '../../tvhelper/dataloaders/EpisodeLoader';
import ProgramsLoader, {
  type ProgramsParams,
  type Programs,
} from '../../trainingjournal/programs/dataloaders/ProgramsLoader';

config();

const { JWT_SECRET } = process.env;

export type LoggedInUser = {|
  +id: string,
  +username: string,
|};

export type GraphqlContextType = {|
  +user: ?LoggedInUser,
  +rawToken: ?string,
  +dataLoader: {|
    +tvhelper: {|
      +searchTvShow: DataLoader<string, TvShow[]>,
      +tvDetail: DataLoader<string, TvShow>,
      +user: DataLoader<string, TvHelperUser>,
      +favorites: DataLoader<string, Favorites[]>,
      +episodes: DataLoader<string, Episode[]>,
      +favorite: DataLoader<FavoriteArgs, Favorites>,
      +episodeWatched: DataLoader<number, EpisodeWatched[]>,
      +episode: DataLoader<string, Episode>,
    |},
    +trainingjournal: {|
      +programs: DataLoader<ProgramsParams, Programs>,
    |},
  |},
|};

const decodeToken = (token: ?string) => {
  try {
    return token == null ? null : jwt.verify(token, JWT_SECRET);
  } catch (err) {
    return null;
  }
};

export default function createContext(token: ?string) {
  const user = decodeToken(token);
  return {
    user,
    rawToken: token,
    dataLoader: {
      tvhelper: {
        searchTvShow: SearchTvShowLoader(),
        tvDetail: TvDetailLoader(),
        user: TvHelperUserLoader(),
        favorites: FavoritesLoader(),
        episodes: TvShowEpisodesLoader(),
        favorite: FavoriteLoader(),
        episodeWatched: EpiosodeWatchedLoader(user),
        episode: EpisodeLoader(),
      },
      trainingjournal: {
        programs: ProgramsLoader(),
      },
    },
  };
}
