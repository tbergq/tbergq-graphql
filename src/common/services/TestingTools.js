// @flow

import { graphql as originalGraphQL } from 'graphql';
import schema from '../../Schema';
import createContext from './GraphqlContext';

export const graphql = async (
  query: string,
  variables: ?Object,
): Promise<Object> =>
  originalGraphQL(schema, query, null, createContext(), variables);
