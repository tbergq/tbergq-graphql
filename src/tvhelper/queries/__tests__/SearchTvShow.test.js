// @flow

import fetch from 'node-fetch';
import { graphql } from '../../../common/services/TestingTools';

it('works', async () => {
  fetch.mockResponse(
    JSON.stringify([
      {
        show: {
          id: 1,
          name: 'lol',
          status: 'status',
          premiered: '2018-11-11',
          rating: { average: 1.1 },
          image: {
            original: 'http://mock.original',
            medium: 'http://mock.medium',
          },
        },
      },
    ]),
  );
  const query = `
    query($query: String!) {
      searchTvShow(query: $query) {
        edges {
          node {
            id
            name
            status
            premiered
            rating
            image {
              medium
              original
            }
          }
        }
      }
    }
  `;
  expect(await graphql(query, { query: 'lol' })).toMatchSnapshot();
});
