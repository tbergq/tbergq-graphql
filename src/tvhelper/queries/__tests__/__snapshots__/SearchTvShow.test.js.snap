// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Object {
  "data": Object {
    "searchTvShow": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "dHZTaG93OjE=",
            "image": Object {
              "medium": "http://mock.medium",
              "original": "http://mock.original",
            },
            "name": "lol",
            "premiered": "2018-11-11",
            "rating": 1.1,
            "status": "status",
          },
        },
      ],
    },
  },
}
`;
